R/nested.R

```diff
@@ -88,3 +88,49 @@ create_filler_nested <- function(pd_nested) {
   select_(pd_nested, ~spaces, ~newlines, ~short, ~everything())
 }
 
+#' Serialize a nested parse table
+#'
+#' Helper function that recursively extracts terminals from a nested tibble.
+#' @param pd_nested A nested parse table.
+#' @param pass_indent Level of indention of a token.
+#' @return A character vector with all tokens in `pd_nested` plus whitespace,
+#'   line break and indention information added between the tokens.
+#' @importFrom purrr pmap
+serialize_parse_data_nested_helper <- function(pd_nested, pass_indent) {
+  out <- pmap(list(pd_nested$terminal, pd_nested$text, pd_nested$child,
+             pd_nested$spaces, pd_nested$newlines, pd_nested$indent),
+           function(terminal, text, child, spaces, newlines, indent) {
+             if (terminal) {
+               c(rep_char(" ", pass_indent), text,
+                 newlines_and_spaces(newlines, spaces))
+             } else {
+               c(serialize_parse_data_nested_helper(child, indent + pass_indent),
+                 newlines_and_spaces(newlines, spaces))
+             }
+        }
+  )
+  out
+}
+
+#' Serialize a nested parse table
+#'
+#' Collapses a nested parse table into its character vector representation. To
+#'   achieve this, obsolete white spaces that were inserted in
+#'   [serialize_parse_data_nested_helper] spaces before the tokens have to be
+#'   removed.
+#' @param pd_nested A nested parse table with line break, spaces and indention
+#'   information.
+#' @return A character string.
+serialize_parse_data_nested <- function(pd_nested) {
+  raw <- serialize_parse_data_nested_helper(pd_nested, pass_indent = 0) %>%
+    unlist()
+  newline <- which(raw == "\n")
+  token <- setdiff(1:length(raw), union(which(raw == ""),
```

Comment:
I think generally every multi-line input for which at least one level of indention is required and there is more than one token per line. Since the indention level is currently added before every token all but the first token end up with too many spaces before them.
Here, I drop the lines 127 to 131 in nested.R, which are

```r
  newline <- which(raw == "\n")
  token <- setdiff(1:length(raw), union(which(raw == ""),
                                        union(grep("^ +$", raw), newline)))
  to_zero <- setdiff(token - 1, newline + 1)
  raw[to_zero] <- ""
```
Using [reprex](https://github.com/tidyverse/reprex) , we get the following example:
``` r
library(magrittr)
indented_multi_line_random <- c(
"            call(       ", 
"  1,", 
"  call2(               ", 
"    2, 3,", "call3(1, 2, 22),", 
"    5", 
"),",
"           144", 
")")
raw <- styler:::compute_parse_data_nested(indented_multi_line_random) %>%
  styler:::create_filler_nested() %>% 
  styler:::indent_round_nested() %>% 
  styler:::serialize_parse_data_nested_helper(pass_indent = 0) %>% 
  unlist()

out <- raw %>% 
  paste0(collapse = "") %>% 
  strsplit("\n", fixed = TRUE) %>% 
  .[[1L]]

out
#> [1] "call("                                             
#> [2] "  1,"                                              
#> [3] "  call2  ("                                        
#> [4] "    2  ,   3  ,"                                   
#> [5] "    call3    (    1    ,     2    ,     22    )  ,"
#> [6] "    5"                                             
#> [7] "  ),"                                              
#> [8] "  144"                                             
#> [9] ")"
cat(out, sep = "\n")
#> call(
#>   1,
#>   call2  (
#>     2  ,   3  ,
#>     call3    (    1    ,     2    ,     22    )  ,
#>     5
#>   ),
#>   144
#> )
```